dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannels.java

```diff
@@ -0,0 +1,165 @@
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.Response;
+import com.palantir.dialogue.core.BalancedScoreTracker.ChannelScoreInfo;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.random.SafeThreadLocalRandom;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+import java.util.List;
+import java.util.Random;
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/** Allows requests for a 'transaction' should land on one host. */
+public final class StickyEndpointChannels {
+
+    private final ImmutableList<? extends EndpointChannelFactory> channels;
+    private final BalancedScoreTracker tracker;
+
+    private StickyEndpointChannels(Builder params) {
+        this.channels = params.channels;
+        // TODO(dfox): handle 'channels' of size 0/1 (balanced tracker will barf)
+        this.tracker = new BalancedScoreTracker(
+                params.channels.size(),
+                params.random,
+                params.ticker,
+                Preconditions.checkNotNull(params.taggedMetrics, "taggedMetricRegistry"),
+                Preconditions.checkNotNull(params.channelName, "channelName"));
+    }
+
+    /**
+     * Returns the current 'best' channel based on its Balanced score. If the underlying host is completely offline,
+     * callers are responsible for getting a new sticky channel by calling this method again, relying on the
+     * {@link BalancedScoreTracker} to avoid returning the same broken channel again.
+     */
+    public Channel getStickyChannel() {
+        return new Sticky(channels, tracker);
+    }
+
+    @ThreadSafe
+    private static final class Sticky implements EndpointChannelFactory, Channel {
+
+        private final ImmutableList<? extends EndpointChannelFactory> channels;
+        private final Supplier<ChannelScoreInfo> getSingleBestChannel;
+
+        private Sticky(ImmutableList<? extends EndpointChannelFactory> channels, BalancedScoreTracker tracker) {
+            this.channels = channels;
+            this.getSingleBestChannel = Suppliers.memoize(tracker::getSingleBestChannelByScore);
+        }
+
+        @Override
+        public EndpointChannel endpoint(Endpoint endpoint) {
+            ChannelScoreInfo channelScoreInfo = getSingleBestChannel.get();
+            int hostIndex = channelScoreInfo.channelIndex();
+            EndpointChannel delegate = channels.get(hostIndex).endpoint(endpoint);
+            return new ScoreTrackingEndpointChannel(delegate, channelScoreInfo);
+        }
+
+        /**
+         * .
+         * @deprecated prefer {@link #endpoint}, as it allows binding work upfront
+         */
+        @Deprecated
+        @Override
+        public ListenableFuture<Response> execute(Endpoint _endpoint, Request _request) {
+            // TODO(dfox): could we delete this entirely?
+            throw new UnsupportedOperationException("Not implemented");
```

Comment:
Since this ends up exposing a Channel/EndpointChannel, I think we should provide the naive implementation of this method to prevent surprises.

Reply:
So I had this originally, but I mistakenly ended up going through _two_ backcompat layers and missing the fast path.  Ideally, I'd like to get conjure-java to codegen a new factory method `.of(EndpointChannelFactory, Runtime)` and then we wouldn't have to pretend everything is a Channel anymore.